### PR TITLE
Update BGP Graceful Restart test requirements

### PR DESCRIPTION
**Pull request by sachendras**
Update RT-1.4 test to verify the BGP graceful restart capability. A router that supports BGP graceful restart can work either as a Restarting speaker mode or in helper mode. By advertising BGP graceful restart capability, a router announces to the peer its ability to,

1. [Restarting Speaker] Maintain forwarding state on all the routes in its FIB even when its BGP process is restarting. Therefore the peer functioning as a helper should continue to direct flows at the subject router undergoing BGP process restart.
2. [Helper Router] Support a peer whose BGP process is restarting by continuing to direct flows at the peer.
3. The test checks support for RFC4724 and RFC8538.
While testing for the above, this test also confirms that the implementation respects stale-routes-time timer setting.

b/316964402 for ATE feature enhancement to delay BGP establishment between ATE and DUT

### CODE REVIEW

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 1]
```diff
@@ -4,33 +4,32 @@
...
-        *   Ensure that prefixes are propagated to ATE port-2 during the
-            restart.
-        *   Ensure that traffic can be forwarded between ATE port-1 and ATE
-            port-2 during stale routes time.
-        *   Ensure that prefixes are withdrawn, and traffic cannot be forwarded
-            between ATE port-1 and port-2 after the stale routes time expires.
+## Procedure
+ *   Configure EBGP peering between ATE:Port1 and DUT:Port1
+ *   Configure IBGP peering between ATE:Port2 and DUT:Port2
+ *   Ensure that the EBGP and IBGP peering are setup for IPv4-Unicast and IPv6-unicast AFI-SAFIs
```
Comment (rszarecki) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:15]: How above are relates to peer-groups? 2 groups, one session each?

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 2]
```diff
@@ -4,33 +4,32 @@
...
+ *   Enable `Graceful-Restart` capability at the `Peer-Group` level.
+ *   Ensure that the `restart-time` and the `stale-routes-time` are configured at the `Global` level
+ *   Configure allow route-policy under BGP peer-group address-family
+ *   Validate received capabilities at DUT and ATE reflect support for graceful
+     restart.
+ *   (Restarting speaker) Advertise prefixes between the ATE ports, through the DUT. Trigger DUT session restart by killing the BGP process in the DUT. Please use the `gNOI.killProcessRequest_Signal_Term` as per [gNOI_proto](https://github.com/openconfig/gnoi/blob/main/system/system.proto#L326).
```
Comment (rszarecki) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:21]:
```suggestion
 *   TestCase - Restarting DUT speaker 
   * Advertise prefixes between the ATE ports, through the DUT. 
   * Trigger DUT session restart by killing the BGP process in the DUT. Please use the `gNOI.killProcessRequest_Signal_Term` as per [gNOI_proto](https://github.com/openconfig/gnoi/blob/main/system/system.proto#L326).
```

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 3]
```diff
@@ -4,33 +4,32 @@
...
+       *   Please kill the right process to restart BGP. For Juniper it is the `RPD` process. For Arista this is the `BGP` process. For Nokia this is `sr_bgp_mgr`. Similar processes need to be included for Cisco.
```
Comment (rszarecki) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:22]:
```suggestion
         *   Please kill the right process to restart BGP. For Juniper it is the `RPD` process. For Arista this is the `BGP` process. For Nokia this is `sr_bgp_mgr`. 
 > TODO: Similar processes need to be included for Cisco.
```

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 4]
```diff
@@ -4,33 +4,32 @@
...
+       *   Once the process is killied, verify that the packets are:
+             *   Forwarded between ATE port-1 and DUT port-1 for the duration of the specified stale routes time.
+             *   Dropped after the stale routes timer has expired.
+             *   Forwarded again between ATE port-1 and DUT port-1 after the session is re-established.
+ *   (Helper for a restarting IBGP speaker) Advertise prefixes between the ATE ports through the DUT. Send Graceful restart trigger from ATE port-2.
```
Comment (rszarecki) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:27]:
```suggestion
 *   TestCase -  DUT Helper for a restarting IBGP speaker
     * Advertise prefixes between the ATE ports through the DUT. Send Graceful restart trigger from ATE port-2.
```

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 5]
```diff
@@ -4,33 +4,32 @@
...
+       *   Ensure that traffic can be forwarded between ATE port-1 and ATE port-2 during stale routes time.
+       *   Ensure that prefixes are withdrawn, and traffic cannot be forwarded between ATE port-1 and port-2 after the stale routes time expires.
+*  (Helper for a restarting EBGP speaker) Advertise prefixes between the ATE ports through the DUT. Send Graceful restart trigger from ATE port-1.
```
Comment (rszarecki) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:30]:
```suggestion
*  TestCase - DUT Helper for a restarting EBGP speaker
    * Advertise prefixes between the ATE ports through the DUT. Send Graceful restart trigger from ATE port-1.
```

**Review by rszarecki — APPROVED** [review]

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 6]
```diff
@@ -4,33 +4,32 @@
 
 BGP Graceful Restart
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:5]: Needs a test ID assigned

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 1 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:15]: Correct, one for EBGP and another for IBGP

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 6 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:5]: This is updating RT-1.4

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 7]
```diff
@@ -4,33 +4,36 @@
...
 
-## Procedure
+## Topology
+Follwing connections:
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:8]:
```suggestion
Create the following connections:
```

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 8]
```diff
@@ -4,33 +4,36 @@
...
+ *   Configure IBGP peering between ATE:Port2 and DUT:Port2
+ *   Ensure that the EBGP and IBGP peering are setup for IPv4-Unicast and IPv6-unicast AFI-SAFIs. Total 2xpeer-groups (1 per protocol) with 1 BGP session each.  
+ *   Enable `Graceful-Restart` capability at the `Peer-Group` level.
+ *   Ensure that the `restart-time` and the `stale-routes-time` are configured at the `Global` level
+ *   Configure allow route-policy under BGP peer-group address-family
+ *   Validate received capabilities at DUT and ATE reflect support for graceful
+     restart.
+ *   TestCase - Restarting DUT speaker 
+   * Advertise prefixes between the ATE ports, through the DUT. 
+   * Trigger DUT session restart by killing the BGP process in the DUT. Please use the `gNOI.killProcessRequest_Signal_Term` as per [gNOI_proto](https://github.com/openconfig/gnoi/blob/main/system/system.proto#L326).
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:23]: Recomend using gNOI for session clear: https://github.com/openconfig/gnoi/blob/main/bgp/bgp.proto#L35

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 9]
```diff
@@ -41,14 +44,16 @@ For prefixes:
...
-*   afi-safis/afi-safi/afi-safi-name
-*   afi-safis/afi-safi/graceful-restart/state/advertised
-*   afi-safis/afi-safi/graceful-restart/state/peer-restart-time
-*   afi-safis/afi-safi/graceful-restart/state/received
+*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/afi-safi-name
+*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/graceful-restart/state/advertised
+*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/graceful-restart/state/peer-restart-time
+*   /network-instances/network-instance/protocols/protocol/bgp/neighbors/neighbor/afi-safis/afi-safi/graceful-restart/state/received
+*   /network-instances/network-instance/protocols/protocol/bgp/global/graceful-restart/state/restart-time
+*   /network-instances/network-instance/protocols/protocol/bgp/global/graceful-restart/state/stale-routes-time
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:59]: Add section for ## gNOI RPC coverage
* gNOI.BGP:ClearBGPNeighborRequest
* gNOI.BGP:ClearBGPNeighborResponse

**Review by dplore — COMMENTED** [review]
Recommend using BGP clear neighbor RPC instead of kill process. I think this will be cleaner.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 8 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:23]: But the goal here is to test BGP GR in the DUT in restarting mode. This requires the BGP process in the DUT to gracefully go down. Clear BGP Neighbor wouldn't do that but [Signal_Term](https://github.com/openconfig/gnoi/blob/main/system/system.proto#L326) would. Please check the conversation in b/275111471

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 9 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:59]: Please check my last comment.

Comment (dplore) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:76]: If the device supports [RFC8538](https://www.rfc-editor.org/rfc/rfc8538.html) then one can use the [BGP reset soft option](https://www.rfc-editor.org/rfc/rfc8538.html#section-4.1) which is also defined in [gnoi.ClearBGPNeighborRequest with mode=SOFT](https://github.com/openconfig/gnoi/blob/main/bgp/bgp.proto#L41) .

If you feel we need to keep a [RFC 4724](https://www.rfc-editor.org/rfc/rfc4724) specific test, that is ok.  RFC8538 compliant behavior, using gnoi.ClearBGPNeighborRequest with mode=SOFT  could be an additional subtest.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 8 above)
Comment (dplore) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md]: Ok, I think I get it now. The kill process `term` method is expected to gracefully restart the BGP process on the DUT.  The DUT is expected to stop it's  BGP TCP session with the ATE, preserve FIB state for BGP learned prefixes and start a new BGP TCP following https://www.rfc-editor.org/rfc/rfc4724.html#section-4.1.  

Seems like we need to ensure traffic is flowing using  prefixes learned from the ATE  during the restart to confirm FIB entries are maintained on the DUT and there is no packet loss during the graceful restart.

Can you add all this text into the test plan?

Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md]: Updated summary section. PTAL

**Review by dplore — DISMISSED** [review]

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 10]
```diff
@@ -2,53 +2,141 @@
...
-        *   Ensure that traffic can be forwarded between ATE port-1 and ATE
-            port-2 during stale routes time.
-        *   Ensure that prefixes are withdrawn, and traffic cannot be forwarded
-            between ATE port-1 and port-2 after the stale routes time expires.
+     restart.
+    
+
+**RT-1.4.2: Restarting DUT speaker whose BGP process was killed gracefully**
+*   Advertise prefixes between the ATE ports, through the DUT. 
+*   Trigger DUT session restart by killing the BGP process in the DUT. Please use the `gNOI.killProcessRequest_Signal_Term` as per [gNOI_proto](https://github.com/openconfig/gnoi/blob/main/system/system.proto#L326).
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:35]: insert step to start traffic

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 11]
```diff
@@ -2,53 +2,141 @@
...
+     *   Please kill the right process to restart BGP. For Juniper it is the `RPD` process. For Arista and Cisco this is the `BGP` process. For Nokia this is `sr_bgp_mgr`.
+     *   Once the process is killied, verify that the packets are:
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:37]: add step to wait for bgp session to be reestablished with ATE

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 12]
```diff
@@ -2,53 +2,141 @@
...
+          *   Forwarded between ATE port-1 and DUT port-1 for the duration of the specified stale routes time.
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:38]: add step to stop traffic.  Verify zero packet loss.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 13]
```diff
@@ -2,53 +2,141 @@
...
+          *   Dropped after the stale routes timer has expired.
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:39]: shouldn't the session recover before the stale routes timer expires?

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md [hunk 14]
```diff
@@ -2,53 +2,141 @@
...
-*   Establish eBGP sessions between:
-    *   ATE port-1 and DUT port-1
-    *   ATE port-2 and DUT port-2
-    *   Configure allow route-policy under BGP peer-group address-family
+**RT-1.4.1: Enable and validate BGP Graceful restart feature**
+*   Configure EBGP peering between ATE:Port1 and DUT:Port1
+*   Configure IBGP peering between ATE:Port2 and DUT:Port2
+*   Ensure that the EBGP and IBGP peering are setup for IPv4-Unicast and IPv6-unicast AFI-SAFIs. Total 2xpeer-groups (1 per protocol) with 1 BGP session each.  
+*   Enable `Graceful-Restart` capability at the `Peer-Group` level.
+*   Ensure that the `restart-time` and the `stale-routes-time` are configured at the `Global` level
```
Comment (dplore) [inline feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:27]: what should stale-routes-time be set to?  (performance target for bgp restart)

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 13 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:39]: Yeah thought about that. We will need to control IXIA from sending BGP packets towards the DUT and manifest delay in the BGP reestablishment process post the kill action. Dont know what options we have.

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 10 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md:36]: done

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 11 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md]: done

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 13 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md]: Added step to address the concern. Also raised [b/316964402](https://buganizer.corp.google.com/issues/316964402) for IXIA support

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 14 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md]: I have defined the numbers for RFC4724 v/s RFC8538

File: feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md (hunk 12 above)
Comment (sachendras) [reply feature/bgp/gracefulrestart/ate_tests/bgp_graceful_restart_test/README.md]: Done

**Review by dplore — APPROVED** [review]

**Pull request merged**